### DISCUSSION TITLE
can't understand the conversions in fundMe - lesson 4

### DISCUSSION THREAD

**Discussion by elec7rik** [Cyfrin/foundry-full-course-cu#727] [Q&A] [answered]
i am not able to understand why are we multiplying - price { while returning in the getPrice() } by 1e10? also why is minUSD = 5e18? aren't we just talking about an amount of 5 USD to be the threshold value for the transaction? 
![fundMe](https://github.com/Cyfrin/foundry-full-course-f23/assets/89316185/ee4c25de-950b-4c97-87b9-b52a8333513b)

**Comment by theirrationalone** [comment, marked as answer]
1. "As we know, the current price of `1 ETH` is equal to around $2000." ✅
2. "We can equate 1000000000000000000 to 1 ETH (1_00000000_0000000000 == 1000000000000000000) in `WEI`."
3. "According to Patrick's intentions, he wants to pay the minimum value in `USD` using `ETH`."
4. "However, ETH is much more expensive, and faucets provide only limited `ETH` (SEPOLIA, GOERLI, etc.) to us (METAMASK)."
5. "So, if we have `0.5 ETH`, for our convenience, Patrick has decided to use only `0.005 ETH` (1% of `0.5 ETH`) so that we can perform at least 100+ transactions on a testnet (i.e., Sepolia)."
6. "Now, let's calculate the value of `0.005 ETH` in USD:
   a. We paid `0.005 ETH`, which in `WEI` would be 5e15 == 5000000000000000 = 5_00000_0000000000 ✅
   b. `1 ETH` = `$2000` -> in `WEI`, it would be 1e18 = 1_00000000_0000000000 = $2000
   c. 0.005 * 2000 -> 5e15 * 2000 -> 5_00000_0000000000 * 2000 -> 5000000000000000 * 2000 = 10_00000000_0000000000 or $10
   d. As we know, if `1 ETH` = $2000, then `0.005 ETH` = $10. But in Solidity, we are limited to integral/decimal values, and floating-point/rational values are not fully supported yet. In programming, there's a concept called overflow/underflow, which occurs when a value crosses the boundary of sufficiency.

# Regarding your questions:
# 1. "Your Answer: Why are we multiplying the price {while returning in getPrice()} by 1e10?"

Now we know that floating-point/rational values are not fully supported yet in Solidity. To avoid underflow/overflow, the Chainlink Oracle decided to provide off-chain price feed data in integral values. Chainlink also provides decimals information to us...
By executing `priceFeed.decimals()`, you can check how many decimals are present in the answer. For example:

If `priceFeed.decimals()` gives an output of 8, and `priceFeed.latestRoundData()` gives an output of 200012304543 (the price, which is the 2nd element in the output tuple).

So, the output of 8 indicates that there is a decimal point after eight digits from the right -> 2000.12304543.

But, this is for our understanding. Now, let's discuss why we multiplied the price by 1e10:

Now we know that there are 8 decimals in `price` -> 200012304543, which is around 2e11. ✅
If we try to perform the division `2e11 / 1e18` to avoid floating-point problems, this division can lead to an underflow. That's why we multiplied the price by 1e10, so the price already has 8 decimals, and we add 10 decimals to make it proportional to 1e18 to avoid underflow.

If we take our example value, it would look like this: 200012304543 * 1e10 -> 200012304543 * 10000000000 -> 2000123045430000000000 -> 2000_12304543_00000000. -> 2000e18.

# 2. "Your answer: Why is minUSD = 5e18? Aren't we just talking about an amount of 5 USD as the threshold value for the transaction?"

First, let's revisit our previous example:

What we have up to now is 2000123045430000000000 or 2000e18. Let's convert the paid prices into USD.

Let's say we paid `0.005 ETH` -> 5e15 -> 5_00000_0000000000 -> 500000000000000.

Conversion: 2000123045430000000000 * 500000000000000 -> 10_061522_7150000000_0000000000_0000000000.

Now, we have a high value -> 10_061522_7150000000_0000000000_0000000000. With this value, if we try to perform division by 1e18, we won't encounter an underflow because of that multiplication by 1e10.

Now, let's perform the division to get our original valid value:

10_061522_7150000000_0000000000_0000000000 / 1_00000000_0000000000 --> 10061522715000000000000000000000000000 / 1000000000000000000 -> 10000615227150000000 -> 10_00061522_7150000000.

It's around $100, as we know 0.005 ETH is equivalent to $10.

And 10e18 is greater than 5e18 (minUSD) -> $10 is way greater than $5. It means that we can pay a small amount of ETH, i.e., `0.003 ETH`, and the USD equivalent value of it will be greater than 5e18.

So in the end, yes, you're correct. The threshold is $5 or 5 USD.

**Reply by elec7rik** [reply]
thanks man it was a really good explanation, really appreciate the effort you put in!
so is it that we eventually end up representing the price(2000 USD for 1 eth) in a form that matches the 18 decimal places of 1e18 because 1 eth = 1e18 wei so this is the standard way of conversion while performing computation?

**Reply by theirrationalone** [reply]
Of course, @tend0pain !, It's a standard way... Wanna play more?? Use this [site](https://eth-converter.com/) as well as your Calculator...
